packages/angular/ssr/src/utils/url.ts

```diff
@@ -128,3 +150,51 @@ export function stripIndexHtmlFromURL(url: URL): URL {
 
   return url;
 }
+
+/**
+ * Resolves dynamic placeholders in a URL template by mapping them to corresponding segments
+ * from a base URL's path.
+ *
+ * This function processes the `urlTemplate` string, replacing any placeholders (e.g., `:param`)
+ * with segments from the `baseUrl`. If the `urlTemplate` contains no placeholders, it resolves
+ * it as a relative URL against the `baseUrl`.
+ *
+ * @param urlTemplate - A string representing the URL template, which may include dynamic placeholders
+ * (e.g., `/:param`) to be replaced with segments from the `baseUrl`'s path.
+ * @param baseUrl - The base URL used as a reference for resolving relative paths and dynamic segments
+ * in the `urlTemplate`. The base URL's pathname is split into parts to map to placeholders.
+ * @returns A fully constructed `URL` object with placeholders replaced, or the resolved relative URL
+ * if no placeholders are present.
+ *
+ * @example
+ * ```typescript
+ * const resolvedUrl = constructUrlWithPlaceholders('/:id/details', new URL('http://example.com/user/123'));
+ * console.log(resolvedUrl.toString()); // Outputs: 'http://example.com/123/details'
+
+ * const staticUrl = constructUrlWithPlaceholders('/static/path', new URL('http://example.com/base'));
+ * console.log(staticUrl.toString()); // Outputs: 'http://example.com/static/path'
+ * ```
+ */
+export function constructUrlWithPlaceholders(urlTemplate: string, baseUrl: URL): URL {
```

Comment:
I think my point is that we've put 4. and 5. under a distinct abstraction (`constructUrlWithPlaceholders`), coupling them together. I was thinking it might make more sense to inline these two steps in `src/app.ts`, but renaming that function to something more meaningful might be fine (`buildRedirectUrl`, `resolveRedirect`, etc.) since that would better define this abstraction.

The other piece is that we are copying URL parameters from one string to another (again, coupling the idea of applying parameters to resolving redirects). I'm thinking that it might be clearer to implement step 5. specifically as applying parameters from an independent object. Whether that's literally:

```typescript
const params = extractParams(baseUrl);
const redirectUrl = applyParams(redirectUrlTemplate, params);
```

in `constructUrlWithPlaceholders` or if we can reuse parameters extracted previously and just apply them to the redirect URL.

This probably isn't super important, my intuition is that the intent would be easier to follow if we apply a given set of parameters to a URL template instead of trying to copy over specific segments from one URL to another. Just something to consider, it might not actually work out better in practice so I'm fine to defer to you in the end.

Reply:
I renamed the method to `buildPathWithParams`. I did try splitting it into two separate methods, but it ended up being more difficult to follow.